### ISSUE TITLE
PC0013 - False positives

### ISSUE THREAD

**Issue by ZepposBE** [ALCops/Analyzers#101]
---

## PC0013 false positive on system tables `AllObj` and `AllObjWithCaption`

**Analyzer:** PlatformCop / PC0013  
**Docs:** https://alcops.dev/docs/analyzers/platformcop/pc0013/

PC0013 is reported on `Record.Get` calls that pass the correct primary key. Both system tables have primary key `("Object Type", "Object ID")`; the two arguments match.

### Example 1 – `AllObj`

```al
var
    Obj: Record AllObj;
    SourceTableID: Integer;
begin
    if not Obj.Get(Obj."Object Type"::Table, SourceTableID) then
        exit;
end;
```

### Example 2 – `AllObjWithCaption`

```al
var
    AllObjWithCaption: Record AllObjWithCaption;
    SourceTableID: Integer;
begin
    if AllObjWithCaption.Get(AllObjWithCaption."Object Type"::Table, SourceTableID) then
        exit(AllObjWithCaption."Object Caption");
end;
```

**Expected:** No PC0013.  
**Actual:** PC0013 is reported; we need `#pragma warning disable/restore PC0013` to silence it.

Likely cause: the analyzer does not resolve the primary key of these system tables correctly.

## Possible cause: (AI analysis)

The PC0013 logic lives in [RecordGetProcedureArguments.cs](https://github.com/ALCops/Analyzers/blob/main/src/ALCops.PlatformCop/Analyzers/RecordGetProcedureArguments.cs). Here’s what’s going on.

---

## Where the false positive comes from

The analyzer does two things:

1. **Resolve the table from the record instance** (lines 66–67):

   ```csharp
   if (invocation.Instance?.Type.GetTypeSymbol()?.OriginalDefinition is not ITableTypeSymbol table)
       return;
   ```

2. **Compare argument count to the table’s primary key field count** (lines 82–95):

   ```csharp
   if (invocation.Arguments.Length != table.PrimaryKey.Fields.Length)
   {
       string expectedArgs = invocation.Arguments.Length < table.PrimaryKey.Fields.Length
           ? $"Insufficient arguments provided; expected {table.PrimaryKey.Fields.Length} arguments"
           : $"Too many arguments provided; expected {table.PrimaryKey.Fields.Length} arguments";
       // ... report PC0013
   }
   ```

So the diagnostic is driven entirely by `table.PrimaryKey.Fields.Length`.

---

## Likely cause of the bug

For **system tables** like `AllObj` and `AllObjWithCaption`, the AL compiler’s symbol model often does not expose the same metadata as for app tables. In particular, `table.PrimaryKey` (or `table.PrimaryKey.Fields`) can be **empty or not fully populated** for these tables.

If `table.PrimaryKey.Fields.Length` is **0** for `AllObj` / `AllObjWithCaption`:

- The code passes 2 arguments: `Object Type` and `Object ID`.
- The check becomes `2 != 0` → true.
- The analyzer reports **“Too many arguments provided; expected 0 arguments”**.

So the issue is not in your AL code; it’s that **the analyzer trusts `table.PrimaryKey.Fields` even when the symbol model leaves it empty for system tables**, and it doesn’t special-case or skip tables with an empty primary key.

---

## What to fix (for a PR or issue)

In **RecordGetProcedureArguments.cs**:

1. **Guard on empty primary key**  
   Before comparing `invocation.Arguments.Length` to `table.PrimaryKey.Fields.Length`, bail out when the table has no primary key fields in the symbol model, e.g.:

   - If `table.PrimaryKey?.Fields` is null or empty, `return` and do not report PC0013 for that call.

2. **Optional: skip known system tables**  
   As a safeguard, you can also skip analysis for tables whose name (or symbol name) is `AllObj` or `AllObjWithCaption` when the primary key is empty, so that valid `Get(Object Type, Object ID)` calls are never flagged.

That way, correct `Record.Get` calls on `AllObj` and `AllObjWithCaption` (and any other system table with an empty primary key in the symbol model) will no longer get a false positive.
